lib/solarwinds_apm/support/otlp_endpoint.rb

```diff
@@ -0,0 +1,203 @@
+# frozen_string_literal: true
+
+# © 2023 SolarWinds Worldwide, LLC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at:http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+module SolarWindsAPM
+  # OTLPEndPoint
+  class OTLPEndPoint
+    SW_ENDPOINT_REGEX = /^apm\.collector\.([a-z0-9-]+)\.cloud\.solarwinds\.com$/
+    OTEL_ENDPOINT_REGEX = %r{^https://otel\.collector\.([a-z0-9-]+)\.cloud\.solarwinds\.com:443(?:/.*)?$}
+    OTEL_ENDPOINT_LOCAL_REGEX = %r{\Ahttp://0\.0\.0\.0:(4317|4318)\z}
+    OTEL_ENDPOINT_LOCAL_REGEX2 = %r{\Ahttp://0\.0\.0\.0:(4317|4318)/v1/(metrics|traces|logs)\z}
+    DEFAULT_OTLP_ENDPOINT = 'https://otel.collector.na-01.cloud.solarwinds.com:443'
+    DEFAULT_APMPROTO_ENDPOINT = 'apm.collector.na-01.cloud.solarwinds.com'
+
+    def initialize
+      @token = nil
+      @service_name = nil
+      @lambda_env = determine_lambda_env
+      @agent_enable = true
+      @localhost = false
+      determine_if_localhost
+    end
+
+    def determine_if_localhost
+      @localhost = true if ENV['OTEL_EXPORTER_OTLP_ENDPOINT'].to_s.match?(OTEL_ENDPOINT_LOCAL_REGEX)
+      @localhost = true if ENV['OTEL_EXPORTER_OTLP_TRACES_ENDPOINT'].to_s.match?(OTEL_ENDPOINT_LOCAL_REGEX2)
+      @localhost = true if ENV['OTEL_EXPORTER_OTLP_METRICS_ENDPOINT'].to_s.match?(OTEL_ENDPOINT_LOCAL_REGEX2)
+      @localhost = true if ENV['OTEL_EXPORTER_OTLP_LOGS_ENDPOINT'].to_s.match?(OTEL_ENDPOINT_LOCAL_REGEX2)
+    end
+
+    def config_otlp_endpoint
+      config_service_name
+      config_token
+      %w[TRACES METRICS LOGS].each { |data_type| configure_otlp_endpoint(data_type) }
+    end
+
+    def config_service_name
+      resource_attributes = ENV['OTEL_RESOURCE_ATTRIBUTES'].to_s.split(',').each_with_object({}) do |resource, hash|
+        key, value = resource.split('=')
+        hash[key] = value
+      end
+
+      @service_name = if @lambda_env
+                        ENV['OTEL_SERVICE_NAME'] || ENV['AWS_LAMBDA_FUNCTION_NAME'] || resource_attributes['service.name']
+                      else
+                        ENV['OTEL_SERVICE_NAME'] || resource_attributes['service.name'] || @service_name || 'None'
+                      end
+
+      ENV['OTEL_SERVICE_NAME'] = @service_name
+    end
```

Comment:
is the idea that this OTLPEndPoint class will centralize service name configuration regardless deployment environment (lambda, non-lambda, etc.)? currently there's the setting of service name for Lambda in the [lambda wrapper](https://github.com/solarwinds/apm-ruby/blob/0595edffca4325f4c2d5b2bfbe01d9185611c088/lambda/otel/layer/wrapper#L7), and also in the [OTelLambdaConfig module](https://github.com/solarwinds/apm-ruby/blob/0595edffca4325f4c2d5b2bfbe01d9185611c088/lib/solarwinds_apm/otel_lambda_config.rb#L26).

Also, the lambea logic should be `OTEL_SERVICE_NAME > OTEL_RESOURCE_ATTRIBUTE (service name) > AWS_LAMBDA_FUNCTION_NAME` instead of resource attr service name being the fallback.

Reply:
thinking more... i'm actually not sure why setting the service name should be part of configuring the OTLP endpoint. there's already code parsing, validating and transforming the [service name](https://github.com/search?q=repo%3Asolarwinds%2Fapm-ruby%20servicekeychecker&type=code) that this logic duplicates/misses. why not just let this class set the OTLP endpoint and let the existing logic take care of setting service name?

Reply:
Removed the service name check logic